src/databricks/labs/ucx/source_code/notebooks/sources.py

```diff
@@ -93,35 +112,156 @@ class NotebookLinter:
+    ):
+        self._context: LinterContext = context
+        self._path_lookup = path_lookup
+        self._session_state = session_state
+        self._notebook: Notebook = notebook
+        # reuse Python linter, which accumulates statements for improved inference
+        self._python_linter: PythonSequentialLinter = cast(PythonSequentialLinter, context.linter(Language.PYTHON))
+        self._python_trees: dict[PythonCell, Tree] = {}  # the original trees to be linted
 
     def lint(self) -> Iterable[Advice]:
+        yield from self._load_tree_from_notebook(self._notebook, True)
         for cell in self._notebook.cells:
-            if isinstance(cell, RunCell):
-                self._load_source_from_run_cell(cell)
             if not self._context.is_supported(cell.language.language):
                 continue
-            linter = self._linter(cell.language.language)
-            for advice in linter.lint(cell.original_code):
+            if isinstance(cell, PythonCell):
+                tree = self._python_trees[cell]
+                advices = self._python_linter.lint_tree(tree)
+            else:
+                linter = self._linter(cell.language.language)
+                advices = linter.lint(cell.original_code)
+            for advice in advices:
                 yield advice.replace(
                     start_line=advice.start_line + cell.original_offset,
                     end_line=advice.end_line + cell.original_offset,
                 )
 
+    def _load_tree_from_notebook(self, notebook: Notebook, register_trees: bool) -> Iterable[Advice]:
+        for cell in notebook.cells:
+            if isinstance(cell, RunCell):
+                yield from self._load_tree_from_run_cell(cell)
+                continue
+            if isinstance(cell, PythonCell):
+                yield from self._load_tree_from_python_cell(cell, register_trees)
+                continue
+
+    def _load_tree_from_python_cell(self, python_cell: PythonCell, register_trees: bool) -> Iterable[Advice]:
+        try:
+            tree = Tree.normalize_and_parse(python_cell.original_code)
+            if register_trees:
+                self._python_trees[python_cell] = tree
+            yield from self._load_children_from_tree(tree)
+        except AstroidSyntaxError as e:
+            yield Failure('syntax-error', str(e), 0, 0, 0, 0)
+
+    def _load_children_from_tree(self, tree: Tree) -> Iterable[Advice]:
+        assert isinstance(tree.node, Module)
+        # look for child notebooks (and sys.path changes that might affect their loading)
+        base_nodes: list[NodeBase] = []
+        base_nodes.extend(DbutilsLinter.list_dbutils_notebook_run_calls(tree))
+        base_nodes.extend(SysPathChange.extract_from_tree(self._session_state, tree))
+        _ = """ run_magics = self._list_run_magics(tree)
+        # magic_commands, _ = MagicCommand.extract_from_tree(tree, self._report_problem)
```

Comment:
what is this? why do we need a string and assign it to unused var?

Reply:
removed, that was in prep of #2091